common/hashing_test.cpp

```diff
@@ -0,0 +1,589 @@
+  EXPECT_THAT(HashValue(reinterpret_cast<void*>(&object1)), Eq(hash1));
+  EXPECT_THAT(HashValue(reinterpret_cast<int*>(&object2)), Eq(hash2));
+  EXPECT_THAT(HashValue(reinterpret_cast<std::string*>(object2.data())),
+              Eq(hash3));
+}
+
+TEST(HashingTest, PairsAndTuples) {
+  // Note that we can't compare hash codes across arity, or in general, compare
+  // hash codes for different types as the type isn't part of the hash.
+  HashCode hash_2zero = HashValue(std::pair(0, 0));
+  EXPECT_THAT(HashValue(std::pair(0, 1)), Ne(hash_2zero));
+  EXPECT_THAT(HashValue(std::pair(1, 0)), Ne(hash_2zero));
+
+  HashCode hash_3zero = HashValue(std::tuple(0, 0, 0));
+  EXPECT_THAT(HashValue(std::tuple(0, 0, 1)), Ne(hash_3zero));
+  EXPECT_THAT(HashValue(std::tuple(0, 1, 0)), Ne(hash_3zero));
+  EXPECT_THAT(HashValue(std::tuple(1, 0, 0)), Ne(hash_3zero));
+
+  // Hashing a 2-tuple and a pair should produce identical results though as we
+  // want to be able to have things like variadic tuple construction be
+  // indistinct from pairs.
+  EXPECT_THAT(HashValue(std::tuple(0, 0)), Eq(hash_2zero));
+  EXPECT_THAT(HashValue(std::tuple(0, 1)), Eq(HashValue(std::pair(0, 1))));
+  EXPECT_THAT(HashValue(std::tuple(1, 0)), Eq(HashValue(std::pair(1, 0))));
+
+  // Integers in tuples should work the same as outside of tuples w.r.t.
+  // converting between different integer types.
+  for (int i : {0, 1, 2, 3, 42}) {
+    SCOPED_TRACE(llvm::formatv("Hashing: ({0}, {0}, {0})", i).str());
+    std::tuple v = {i, i, i};
+    HashCode hash = HashValue(v);
+
+    // Zero should match, and other integers shouldn't collide trivially.
+    if (i == 0) {
+      EXPECT_THAT(hash, Eq(hash_3zero));
+    } else {
+      EXPECT_THAT(hash, Ne(hash_3zero));
+    }
+
+    // We shouldn't include the exact integer type used so that implicit
+    // conversions don't shift the hash for non-negative integers, making all of
+    // these match.
+    int8_t i_i8 = i;
+    EXPECT_THAT(HashValue(std::tuple(i_i8, i_i8, i_i8)), Eq(hash));
+    uint8_t i_u8 = i;
+    EXPECT_THAT(HashValue(std::tuple(i_u8, i_u8, i_u8)), Eq(hash));
+    int16_t i_i16 = i;
+    EXPECT_THAT(HashValue(std::tuple(i_i16, i_i16, i_i16)), Eq(hash));
+    uint16_t i_u16 = i;
+    EXPECT_THAT(HashValue(std::tuple(i_u16, i_u16, i_u16)), Eq(hash));
+    int32_t i_i32 = i;
+    EXPECT_THAT(HashValue(std::tuple(i_i32, i_i32, i_i32)), Eq(hash));
+    uint32_t i_u32 = i;
+    EXPECT_THAT(HashValue(std::tuple(i_u32, i_u32, i_u32)), Eq(hash));
+    int64_t i_i64 = i;
+    EXPECT_THAT(HashValue(std::tuple(i_i64, i_i64, i_i64)), Eq(hash));
+    uint64_t i_u64 = i;
+    EXPECT_THAT(HashValue(std::tuple(i_u64, i_u64, i_u64)), Eq(hash));
+
+    // Heterogeneous integer types should also work.
+    EXPECT_THAT(HashValue(std::tuple(i_i8, i_u32, i_i16)), Eq(hash));
+    EXPECT_THAT(HashValue(std::tuple(i_u32, i_i16, i_u64)), Eq(hash));
+  }
+
+  // Pointers inside pairs and tuples should also work like pointers outside.
```

Comment:
"work like pointers outside" is a bit unspecific, I don't know what you mean. I'm guessing this is about the type of the pointer doesn't matter?

This also seems to be testing that a pair of pointers acts like a 2-tuple of pointers, which is something beyond "working like pointers outside".

Maybe:
```suggestion
  // Hash values of pointers in pairs and tuples reflect the address and not the type.
  // Pairs and 2-tuples give the same hash values.
```